eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs

```diff
@@ -0,0 +1,629 @@
+    , networkId = Testnet
+    , systemStart = SystemStart $ posixSecondsToUTCTime 0
+    }
+
+runShelleyBase :: ShelleyBase a -> a
+runShelleyBase act = runIdentity $ runReaderT act testGlobals
+
+submitTx_ ::
+  forall era.
+  ( EraGov era
+  , State (EraRule "LEDGER" era) ~ LedgerState era
+  , BaseM (EraRule "LEDGER" era) ~ ReaderT Globals Identity
+  , Signal (EraRule "LEDGER" era) ~ Tx era
+  , Environment (EraRule "LEDGER" era) ~ LedgerEnv era
+  , STS (EraRule "LEDGER" era)
+  ) =>
+  Tx era ->
+  ImpTestM
+    era
+    ( Either
+        [PredicateFailure (EraRule "LEDGER" era)]
+        (State (EraRule "LEDGER" era))
+    )
+submitTx_ tx = do
+  ImpTestState {..} <- get
+  let
+    ls = impNES ^. nesEsL . esLStateL
+    pp = ls ^. lsUTxOStateL . utxosGovStateL . curPParamsGovStateL
+    trc =
+      TRC
+        ( LedgerEnv 0 minBound pp def
+        , ls
+        , tx
+        )
+  pure $ runShelleyBase (applySTSTest @(EraRule "LEDGER" era) trc)
+
+-- | Submit a transaction that is expected to be accepted. The inputs and
+-- outputs are automatically balanced.
+submitTx ::
+  ( HasCallStack
+  , EraGov era
+  , EraTx era
+  , State (EraRule "LEDGER" era) ~ LedgerState era
+  , ToExpr (PredicateFailure (EraRule "LEDGER" era))
+  , BaseM (EraRule "LEDGER" era) ~ ReaderT Globals Identity
+  , Environment (EraRule "LEDGER" era) ~ LedgerEnv era
+  , Signal (EraRule "LEDGER" era) ~ Tx era
+  , STS (EraRule "LEDGER" era)
+  ) =>
+  String ->
+  Tx era ->
+  ImpTestM era (TxId (EraCrypto era))
+submitTx msg tx = do
+  res <- submitTx_ tx
+  case res of
+    Right x -> do
+      modify $ impNESL . nesEsL . esLStateL .~ x
+      pure . TxId $ hashAnnotated (tx ^. bodyTxL)
+    Left es ->
+      impIO . error $
+        "Failed to submit transaction "
+          <> show msg
+          <> ":\n"
+          <> unlines (showExpr <$> es)
+
+-- | Submit a transaction that is expected to be rejected. The inputs and
+-- outputs are automatically balanced.
+submitFailingTx ::
+  ( HasCallStack
+  , EraGov era
+  , State (EraRule "LEDGER" era) ~ LedgerState era
+  , BaseM (EraRule "LEDGER" era) ~ ReaderT Globals Identity
+  , Signal (EraRule "LEDGER" era) ~ Tx era
+  , Environment (EraRule "LEDGER" era) ~ LedgerEnv era
+  , STS (EraRule "LEDGER" era)
+  ) =>
+  Tx era ->
+  ImpTestM era ()
+submitFailingTx tx = do
+  res <- submitTx_ tx
+  case res of
+    Right _ -> error "Expected transaction to fail, but it succeeded"
+    Left [] -> error "Failed with no errors"
+    Left _ -> pure ()
+
+waitTick ::
+  forall era.
+  ( HasCallStack
+  , BaseM (EraRule "TICK" era) ~ ReaderT Globals Identity
+  , Environment (EraRule "TICK" era) ~ ()
+  , State (EraRule "TICK" era) ~ NewEpochState era
+  , STS (EraRule "TICK" era)
+  , Signal (EraRule "TICK" era) ~ SlotNo
+  , EraTxOut era
+  , NFData (GovState era)
+  , NFData (StashedAVVMAddresses era)
+  , NFData (PredicateFailure (EraRule "TICK" era))
+  ) =>
+  ImpTestM era ()
+waitTick = do
```

Comment:
I don't really like `runTICK`, because we're using `runLEDGER` and `runEPOCH` in another test suite and these work a bit differently from this one. Maybe `passTick` would be better since we already have `passEpoch`?

Reply:
Yeah, `passTick` sounds good :+1: